BB2-823: Attempt to fix docker issues

bfausett77 opened the pull request:
This is more of a PR to review the entire 'Server' side of this application.  To get this to run follow the README.

Let me know if you have questions, issues, suggestions.

nbragdon commented:
@bfausett77 I think the readme could use a little work, here are a couple of things I would add:

- One line saying something like "Setup Docker" which is just a link to docker documentation on getting it setup
- After the docker compose, we should include where they should go to see the app running: http://localhost:3000/
- We should include a link (probably before config) to tell them how to create a sandbox account + get the credentials
- We should also tell them what callback uri they should register for their app, in order for the sample app to work

bfausett77 commented:
> @bfausett77 I think the readme could use a little work, here are a couple of things I would add:
> 
> * One line saying something like "Setup Docker" which is just a link to docker documentation on getting it setup
> * After the docker compose, we should include where they should go to see the app running: http://localhost:3000/
> * We should include a link (probably before config) to tell them how to create a sandbox account + get the credentials

Excellent suggestions.  I've updated the README accordingly.

dtisza1 reviewed (changes requested):
@bfausett77 I was able to get the client & server to startup OK.  However, I'm not seeing any data in the app.  From the docker-compose logging, it looks OK to me, but not seeing data returned.

Also, you can copy some of the python related configs from the BB2 .gitignore file so that "*.pyc" and other files don't get included in the repo.

LINK:  https://github.com/CMSgov/bluebutton-web-server/blob/master/.gitignore

For example, I think everything could be included up to the "# Djano" comment and maybe some other items.

nbragdon commented:
@dtisza1 I had the same thing happen on mine... @bfausett77 i'm wondering if you setup something locally that isn't checked in that we didn't setup on ours?

bfausett77 commented:
> @bfausett77 I was able to get the client & server to startup OK. However, I'm not seeing any data in the app. From the docker-compose logging, it looks OK to me, but not seeing data returned.
> 
> Also, you can copy some of the python related configs from the BB2 .gitignore file so that "*.pyc" and other files don't get included in the repo.
> 
> LINK: https://github.com/CMSgov/bluebutton-web-server/blob/master/.gitignore
> 
> For example, I think everything could be included up to the "# Djano" comment and maybe some other items.

I have updated the .gitignore file and removed the folders/files that shouldn't be there.

bfausett77 commented:
> @dtisza1 I had the same thing happen on mine... @bfausett77 i'm wondering if you setup something locally that isn't checked in that we didn't setup on ours?

Worked with you to find the issue in the views.py file - it was the if-else statement that was the issue

dtisza1 commented:
@bfausett77 Just tried it after your last updates.  Still having an issue.

It looks related to the callback and exchanging the code for an access token with the /v2/o/token/ endpoint (gets a 400 response code per the Splunk logging.

My Sandbox app has this added to the redirect URI's list, which I believe is good:  "http://localhost:3001/api/bluebutton/callback/"

Below are the docker-compose logs:

```
server_1  | HERE AUTHURLhttps://sandbox.bluebutton.cms.gov/v2/o/authorize?client_id=AdGJlRUrAl06ooDnMhNPccwcLYbLEu7ssRIRDrvE&redirect_uri=http://localhost:3001/api/bluebutton/callback/&state=fmpUJi5JUFd9NzdmU3AickU4RDc0VGQiWixrPFBIMGE=&response_type=code&code_challenge_method=S256&code_challenge=3PO7tAsg4NIs0F_jDJZuD3Kojs5sCuubq37DocIq-2A=
server_1  | 2021-11-09 15:52:49,811:werkzeug:INFO:172.30.0.3 - - [09/Nov/2021 15:52:49] "GET /api/authorize/authurl?pkce=true&version=v2&env=sandbox HTTP/1.1" 200 -
server_1  | HERE callback:
server_1  | code
server_1  | dD0iQ2mqrJLtNpUU56lbRzffqQZ53h
server_1  | state
server_1  | fmpUJi5JUFd9NzdmU3AickU4RDc0VGQiWixrPFBIMGE=
server_1  | configsettings
server_1  | {'bb2BaseUrl': 'https://sandbox.bluebutton.cms.gov', 'bb2ClientId': 'AdGJlRUrAl06ooDnMhNPccwcLYbLEu7ssRIRDrvE', 'bb2ClientSecret': 'ZZZZZZ-REMOVED--ZZZZ', 'bb2CallbackUrl': 'http://localhost:3001/api/bluebutton/callback/', 'port': 3001, 'host': 'server'}
server_1  | settings
server_1  | <src.entities.Settings.Settings object at 0x7f8c999fb880>
server_1  | authtoken:
server_1  | {'error': 'invalid_request', 'error_description': 'Mismatching redirect URI.'}
server_1  | logged User AuthTOken
server_1  | {'error': 'invalid_request', 'error_description': 'Mismatching redirect URI.'}
server_1  | AuthToken Access Token
server_1  | None
server_1  | BB2 BENEFIT URL
server_1  | https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit/
server_1  | ERROR:
server_1  | can only concatenate str (not "NoneType") to str
server_1  | 2021-11-09 15:52:59,712:werkzeug:INFO:172.30.0.1 - - [09/Nov/2021 15:52:59] "GET /api/bluebutton/callback/?code=dD0iQ2mqrJLtNpUU56lbRzffqQZ53h&state=fmpUJi5JUFd9NzdmU3AickU4RDc0VGQiWixrPFBIMGE%3D HTTP/1.1" 302 -
server_1  | 2021-11-09 15:52:59,873:werkzeug:INFO:172.30.0.3 - - [09/Nov/2021 15:52:59] "GET /api/data/benefit HTTP/1.1" 200 -

```

nbragdon commented on server/src/app/views.py:
```diff
@@ -72,6 +76,14 @@ def authorizationCallback():
         configSettings = getConfigSettings(myEnv)
 
         # this gets the token from Medicare.gov once the 'user' authenticates their Medicare.gov account
+        print('code')
```
We'll probably want to remove these before we merge this in

bfausett77 replied:
This has been completed

nbragdon commented:
@bfausett77 Where did you get the logger factory class from? Also looks like it has some commented out lines

nbragdon commented:
@bfausett77 Looking around I noticed quite a few commented out items in different files - maybe take a clean up pass through?

nbragdon commented:
@bfausett77 All the changes helped! It's now working for me

bfausett77 commented:
> @dtisza1 I had the same thing happen on mine... @bfausett77 i'm wondering if you setup something locally that isn't checked in that we didn't setup on ours?

This should now be resolved as the changes I made yesterday

bfausett77 commented:
> @bfausett77 Looking around I noticed quite a few commented out items in different files - maybe take a clean up pass through?

I just completed this.  Take a look and let e know if you have any other suggestions.

nbragdon commented on .gitignore:
```diff
@@ -32,3 +26,44 @@ server/src/entities/__pycache__/Settings.cpython-39.pyc
+*$py.class
+
+# Distribution / packaging
+.Python
```
This ignore list seems big! Is there any way to do it more top level? Like an entire directory?

bfausett77 replied:
It is large and we may be able to reduce it.  I will tackle this with any new functionality that we add.

nbragdon commented on client/package.json:
```diff
@@ -1,52 +1,52 @@
+    "name": "my-app",
+    "version": "0.1.0",
+    "private": true,
+    "proxy": "http://server:3001",
```
Do we need this proxy here?

bfausett77 replied:
It appears so.

nbragdon commented on server/src/app/views.py:
```diff
@@ -125,7 +129,9 @@ def authorizationCallback():
     else:
-        return json.loads(loggedInUser.get('eobData'))
\ No newline at end of file
+        return ''
```
Would it make sense to return an empty object here instead?

bfausett77 replied:
I was doing that before, but then the parsing on the front-end failed straight out.  We should, in the next iteration, tackle how to handle not returning any results on the front-end.

nbragdon reviewed (approved):
Left some non-required things to think about in the comments. Take a look and if you want to change them 👍 , otherwise it should be fine merging this in.

bfausett77 commented:
> @bfausett77 Where did you get the logger factory class from? Also looks like it has some commented out lines

These should all be cleaned up now.

dtisza1 commented on .gitignore:
```diff
@@ -32,3 +26,44 @@ server/src/entities/__pycache__/Settings.cpython-39.pyc
 server/src/prestart/__pycache__/setenv.cpython-39.pyc
 server/src/shared/__pycache__/LoggerFactory.cpython-39.pyc
 server/src/utils/__pycache__/generatePKCE.cpython-39.pyc
```
@bfausett77 After adding "*.pyc", these specific .pyc files can be removed from this file.

bfausett77 replied:
Done

dtisza1 reviewed (changes requested):
@bfausett77 I just tested and not having that callback URI placement issue anymore. It's working as expected.

This is looking good to me, except that .pyc files should be removed from this repo.

While looking through the files, I put this list and command line together that will remove them below:

```
git rm server/src/__pycache__/__init__.cpython-39.pyc server/src/app/__pycache__/__init__.cpython-39.pyc server/src/configs/__pycache__/__init__.cpython-39.pyc server/src/configs/__pycache__/config.cpython-39.pyc server/src/data/__pycache__/Database.cpython-39.pyc
 server/src/data/__pycache__/User.cpython-39.pyc server/src/data/__pycache__/__init__.cpython-39.pyc server/src/entities/__pycache__/AuthorizationToken.cpython-39.pyc server/src/entities/__pycache__/Settings.cpython-39.pyc server/src/entities/__pycache__/__init__.cpython-39.pyc server/src/prestart/__pycache__/__init__.cpython-39.pyc server/src/prestart/__pycache__/setenv.cpython-39.pyc server/src/shared/__pycache__/LoggerFactory.cpython-39.pyc server/src/shared/__pycache__/__init__.cpython-39.pyc server/src/utils/__pycache__/__init__.cpython-39.pyc server/src/utils/__pycache__/bb2Util.cpython-39.pyc server/src/utils/__pycache__/configUtil.cpython-39.pyc server/src/utils/__pycache__/generatePKCE.cpython-39.pyc server/src/utils/__pycache__/userUtil.cpython-39.pyc

# Then commit and push up changes.
```

dtisza1 reviewed (approved):
@bfausett77 This looks good to me!  Great job on this!
